packages/services/src/contents/index.ts

```diff
@@ -1576,3 +1625,236 @@ namespace Private {
     return extension;
   }
 }
+
+class ContentProviderRegistry implements IContentProviderRegistry {
+  constructor(options: ContentProviderRegistry.IOptions) {
+    this.register('default', options.defaultProvider);
+    this._defaultProvider = options.defaultProvider;
+  }
+
+  register(identifier: string, provider: IContentProvider): IDisposable {
+    if (this._providers.has(identifier)) {
+      throw Error(
+        `Provider with ${identifier} identifier was already registered on this drive`
+      );
+    }
+    this._providers.set(identifier, provider);
+
+    const fileChangedProxy = (
+      _provider: IContentProvider,
+      change: Contents.IChangedArgs
+    ) => {
+      this._fileChanged.emit(change);
+    };
+    if (provider.fileChanged) {
+      provider.fileChanged.connect(fileChangedProxy);
+    }
+
+    return new DisposableDelegate(() => {
+      if (provider.fileChanged) {
+        provider.fileChanged.disconnect(fileChangedProxy);
+      }
+
+      if (this._providers.has(identifier)) {
+        this._providers.delete(identifier);
+      }
+    });
+  }
+
+  getProvider(identifier?: string): IContentProvider {
+    if (!identifier) {
+      return this._defaultProvider;
+    }
+    const provider = this._providers.get(identifier);
+    if (!provider) {
+      throw Error(`Provider ${identifier} is not registered`);
+    }
+    return provider;
+  }
+
+  get fileChanged(): ISignal<IContentProviderRegistry, Contents.IChangedArgs> {
+    return this._fileChanged;
+  }
+
+  private _providers: Map<string, IContentProvider> = new Map();
+  private _defaultProvider: IContentProvider;
+  private _fileChanged = new Signal<
+    IContentProviderRegistry,
+    Contents.IChangedArgs
+  >(this);
+}
+
+namespace ContentProviderRegistry {
+  /**
+   * Initialization options for `ContentProviderRegistry`.
+   */
+  export interface IOptions {
+    /**
+     * Default provider for the registry.
+     */
+    defaultProvider: IContentProvider;
+  }
+}
+
+/**
+ * A content provider using the Jupyter REST API.
+ */
+export class RestContentProvider implements IContentProvider {
+  identifier: 'rest';
```

Comment:
Is this used anywhere?

Reply:
Good catch - nope, it is not.

Reply:
Resolved by https://github.com/jupyterlab/jupyterlab/pull/17092/commits/ae36421391c8302c38c0dde44e057220b44da2fa.